src/providers/microsoft/azure/mod.rs

```diff
@@ -412,7 +412,22 @@ impl MetadataProvider for Azure {
             bail!("unexpected empty certificates endpoint");
         }
 
-        let key = self.get_ssh_pubkey(certs_endpoint)?;
+        // The certs endpoint is being populated even when the user does not supply
+        // an SSH key when creating a VM. In this case the certificate chain is empty
+        // causing a failure when trying to access it. Check if the certificate chain
+        // is empty, meaning that an SSH key hasn't been provided, and only warn
+        // instead of fail.
+        let key = match self.get_ssh_pubkey(certs_endpoint) {
+            Ok(k) => k,
+            Err(e) => match e.root_cause().to_string().as_ref() {
+                "detected empty certificate chain" => {
```

Comment:
So instead of string matching errors, a more rustic way to do this is to have `get_ssh_pubkey()` return a `Result<Option<PublicKey>>`. Then, `get_ssh_pubkey()` would return `Ok(None)` for this scenario. And here, we'd do e.g.:

```rust
let key: Vec<PublicKey> = self.get_ssh_pubkeys(certs_endpoint)?.unwrap_or_default();
```

Reply:
I modified the line based on your suggestion. Since `PublicKey` doesn't implement Default, I used `unwrap_or(vec![])` instead to assign an empty vector if a None value is returned.